dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,679 @@
+	// a single, empty request is used.
+	//
+	// +listType=atomic
+	DefaultRequests []ResourceRequest `json:"defaultRequests" protobuf:"bytes,5,name=requests"`
 }
 
-// DeviceClassSpec provides either a selector to find other
-// classes, or the specific constraints defining this class.
-type DeviceClassSpec struct {
+type ResourceClassClaimOptions struct {
+	// Config defines configuration parameters that apply to each claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,3,opt,name=config"`
 
-	// One of selector and classCriteria must be populated.
+	// Constraints are additional restrictions for the combination of devices that get
+	// allocated for a claim referencing this class.
+	//
+	// +optional
+	// +listType=atomic
+	Constraints []Constraint `json:"constraints,omitempty"`
+}
+
+type ResourceClassRequestOptions struct {
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty"`
 
-	// Selector specifies a label selector used to identify classes which
-	// may be considered equivalent to this class. In other words, this
-	// allows the adminstrator to define groups of classes which may be
-	// referred to as a single class in the claim.
+	// Requirements must be satisfied by devices. Applies to all devices of
+	// a claim when the claim references the class and only to the devices
+	// in a request when referenced there.
 	//
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
 
-	// Definition specifies the criteria to determine if a device is part
-	// of this class.
-	Definition *DeviceClassDefinition `json:"definition,omitempty"`
+// Constraint must have one and only one field set.
+type Constraint struct {
+	// The match criteria must be satisfied before a set of devices will be used
+	// together.
+	//
+	// +optional
+	Match *MatchModel `json:"match,omitempty"`
 }
 
-// DeviceClassDefinition specifies the subset of devices to consider part of
-// this class, along with device-specific configuration information for those
-// devices.
-type DeviceClassDefinition struct {
-	// Driver specifies the driver that should handle this class of devices.
-	// When a DeviceClaim uses this class, only devices published by the
-	// specified driver will be considered.
-	// +required
-	Driver string `json:driver,omitempty`
+// ConfigurationParameters must have one and only one field set.
+type ConfigurationParameters struct {
+	Vendor *VendorConfigurationParameters `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
```

Comment:
We replaced completely opaque parameters with opaque config + structured selection parameters. I can imagine that at some point in the future we will also define in-tree configuration parameters for certain standardized classes of hardware. Those parameters then would get added here.

Reply:
I'm not sure I buy it oer that it would be worse to just up-scope these.

IOW, You're doing:

```
Class {
  config:
  - vendor:
      parameters: ...
    otherKind:
      ...
}
```

vs.

```
Class {
  config:
    parameters: ...
  otherConfig:
    ...
}
```

I'm not SURE it's better, but one less indirection...

Reply:
Remember that `parameters` would be a list (https://github.com/kubernetes-sigs/wg-device-management/pull/14/files#r1616366220), just like this potential alternative. I prefer to make the list the common outer field under `config` and keep the one-of atomic (no list).

Reply:
List does matter, but still, I need to write it out...

As proposed:
```
Class {
    config:
    - vendor:
      driver: example.com/foo
      parameters: ...
    - vendor:
      driver: example.com/bar
      parameters: ...
    - other:
        something: ...
        somethingElse: ...
```

Right?

I guess am saying:

```
Class {
    config:
      vendor:
      - driver: example.com/foo
        parameters: ...
      - driver: example.com/bar
        parameters: ...
      other:
        something: ...
        somethingElse: ...
```

Isn't that a little cleaner?

Reply:
What is the "one-of" here?

If it's "config", then we cannot have both "vendor" and "other" in the same class, which could be useful.

If it's "other", then we need to pick a proper name for it and define it as an empty struct. Not nice?

Reply:
So in your proposal there is no one-of. "config" is a struct with (currently) a single "vendor" field, which contains a list. I can be convinced to switch to that.

I've not done it yet because I have two (weak?) arguments in favor of the current structure besides the fail-safe future extensions:
- It's consistent with other lists (constraints, requests).
- When combining vendor-defined configuration and Kubernetes-defined configuration, the order might matter.

Reply:
I don't care a ton about this, really.  I can't motivate myself to argue more strongly.  I'm fine with it as-is, except I don't care for the word "vendor".  Can I argue for "opaque" or something like that?

FTR: 

I don't believe there's a need for more than 1 config per-driver, but it doesn't hurt to allow it (actually it makes simpler validation).

I don't believe there's a need for ordering, but it doesn't hurt to allow it.

I don't even believe there will ever be non-external config, but it doesn't hurt to allow it.

Reply:
"opaque" is fine with me.